macros/calendar_date/string_to_date.sql

```diff
@@ -0,0 +1,152 @@
+{%- macro string_to_date(date_string, format='yyyy-mm-dd') -%}
+    {{ adapter.dispatch('string_to_date', 'dbt_date') (date_string, format) }}
+{%- endmacro -%}
+
+{%- macro default__string_to_date(date_string, format) -%}
+-- redshift and postgres are actually pretty good at detecting date strings. we'll add compatibility for Day/Month/Year format
+
+{% if format|lower in ('d/m/yy', 'dd/mm/yy') %}
+-- Ex: (0)1/(4)/22
+to_date('{{ date_string }}', 'dd/mm/yy')
+
+{% elif format|lower in ('d/m/yyyy', 'dd/mm/yyyy') %}
+-- Ex: (0)1/(4)/2022
+to_date('{{ date_string }}', 'dd/mm/yyyy')
+
+{% else %}
+cast('{{ date_string }}' as date)
+
+{% endif %}
+
+{%- endmacro -%}
+
+{%- macro bigquery__string_to_date(date_string, format) -%}
+
+{% if format|lower[:3] = 'mon' %}
+-- Ex: Apr(il) (0)1, 2022
+parse_date("%b %e, %Y", '{{ date_string }}')
+
+{% elif format|lower[:5] = 'd mon' or format|lower[:6] = 'dd mon' %}
+-- Ex: (0)1 Apr(il) 2022
+parse_date("%e %b %Y", '{{ date_string }}')
+
+{% elif format|lower in ('m/d/yy', 'mm/dd/yy') %}
+-- Ex: (0)4/(0)1/22
+parse_date("%D", '{{ date_string }}')
+
+{% elif format|lower in ('m/d/yyyy', 'mm/dd/yyyy') %}
+-- Ex: (0)4/(0)1/2022
+parse_date("%m/%e/%Y", '{{ date_string }}')
+
+{% elif format|lower in ('d/m/yy', 'dd/mm/yy') %}
+-- Ex: (0)1/(4)/22
+parse_date("%e/%m/%y", '{{ date_string }}')
+
+{% elif format|lower in ('d/m/yyyy', 'dd/mm/yyyy') %}
+-- Ex: (0)1/(4)/2022
+parse_date("%e/%m/%Y", '{{ date_string }}')
+
+{% elif format|lower in ('dd-mon-yyyy', 'd-mon-yyyy') %}
+-- Ex: (0)1-APR-2022
+parse_date("%e-%b-%Y", '{{ date_string }}')
+
+{% elif format|lower in ('dd-mon-yyyy', 'd-mon-yyyy') %}
+-- Ex: APR-(0)1-2022
+parse_date("%b-%e-%Y", '{{ date_string }}')
+
+{% else %}
+parse_date("%F", '{{ date_string }}')
+
+{% endif %}
+
+{%- endmacro -%}
+
+{%- macro snowflake__string_to_date(date_string, format) -%}
+
+{% if format|lower[:3] = 'mon' %}
```

Comment:
Is there any way to make these string slicing calls more robust? I find myself getting lost in what we're slicing. I fear some of these longer `if`s are hard to debug to figure out where in the stack the error is (since we're depending on the fall-through ordering of the `if` cases, right?)